developer_manual/app_publishing_maintenance/release_automation.rst

```diff
@@ -0,0 +1,197 @@
+Release Automation
+==================
+
+.. sectionauthor:: Benjamin Brahmer <info@b-brahmer.de>
+
+Automation is a great thing, it prevents mistakes and makes your life easier, which gives you more time to work on all those features you wanted to implement.
+
+GitHub Actions
+--------------
+If your application's repository lives on GitHub, as many Nextcloud applications do, GitHub Actions is a great way to automate the release of your app from the git repository into the Nextcloud App Store.
+
+One easy way to get you started is to use https://github.com/R0Wi/nextcloud-appstore-push-action in your repository together with a few other actions. You can automatically build your app and publish it to the App Store. It supports pre-releases and code signing.
+To get started you create a new yaml file in the ``.github/workflows`` directory.
+
+.. code-block:: yaml
+
+    name: Build and publish app release
+
+    on:
+      release:
+          types: [published]
+
+    env:
+      APP_NAME: news
+
+    jobs:
+      build_and_publish:
+        environment: release
+        runs-on: ubuntu-latest
+        name: "Release: build, sign and upload the app"
+        strategy:
+          matrix:
+            php-versions: ['7.4']
+            nextcloud: ['stable21']
+            database: ['sqlite']
+        steps:
+          - name: Checkout
+            uses: actions/checkout@v2
+
+          - name: Setup PHP
+            uses: shivammathur/setup-php@v2
+            with:
+              php-version: ${{ matrix.php-versions }}
+              extensions: pdo_sqlite,pdo_mysql,pdo_pgsql,gd,zip
+              coverage: none
+
+          - name: Set up server non MySQL
+            uses: SMillerDev/nextcloud-actions/setup-nextcloud@main
+            with:
+              cron: false
+              version: ${{ matrix.nextcloud }}
+              database-type: ${{ matrix.database }}
+
+          - name: Prime app build
+            run: make
+
+          - name: Configure server with app
+            uses: SMillerDev/nextcloud-actions/setup-nextcloud-app@main
+            with:
+              app: ${{ env.APP_NAME }}
+              check-code: false
+
+          - name: Create signed release archive
+            run: cd ../server/apps/${{ env.APP_NAME }} && make appstore
+            env:
+              app_private_key: ${{ secrets.APP_PRIVATE_KEY }}
+              app_public_crt: ${{ secrets.APP_PUBLIC_CRT }}
+
+          - name: Upload app tarball to release
+            uses: svenstaro/upload-release-action@v2
+            id: attach_to_release
+            with:
+              repo_token: ${{ secrets.GITHUB_TOKEN }}
+              file: ../server/apps/${{ env.APP_NAME }}/build/artifacts/appstore/${{ env.APP_NAME }}.tar.gz
+              asset_name: ${{ env.APP_NAME }}.tar.gz
+              tag: ${{ github.ref }}
+              overwrite: true
+
+          - name: Upload app to Nextcloud appstore
+            uses: R0Wi/nextcloud-appstore-push-action@v1
```

Comment:
How does version pinning work? What happens if a malicious `v1.0.4` is released or such? Similar for the other versions above and the references to `main`.

Reply:
What do you mean by version pinning?

We have two areas were an app can be "released", we have GitHub and the App Store.
Anyone that is part of the nextcloud organization or got dedicated write access to a repository within nextcloud can create a new release on GitHub, a new release automatically triggers the creation of a tag and to my understanding there is nothing we can do about that.

So let's say one approval is required for app X to merge into main, User A commits the malicious code, User B reviews and approves it, the code gets merged. User A continues to release a new version via GitHub, a new tag is created at the same time, we end up with a release on GitHub without the `app.tar.gz`

For the run of this workflow as you find it in the documentation an approval is required. Anyone that is an admin of the nextcloud organization or the repository can change the approval requirements. There is no protection against a rouge admin in GitHub apart from audit logs, which I can't access. The approval is handled via GitHub currently either directly in the action or via the beta notification system, things might change though.

If someone prefers to host their app in a private account the organization part of course can be ignored and it only matters who gains write access to the repository.

Reply:
https://github.com/nextcloud/documentation/blob/dc3fe5d805244f04d013e414b96a3ce5b5912a50/developer_manual/app_publishing_maintenance/release_automation.rst#the-process